tests/test_indexing.py

```diff
@@ -1995,6 +1998,71 @@ def test_iter_chunk_regions():
         assert_array_equal(a[region], np.zeros_like(a[region]))
 
 
+@pytest.mark.parametrize(
+    ("domain_shape", "region_shape", "origin", "selection_shape"),
+    [
+        ((9,), (1,), None, (9,)),
+        ((9,), (1,), (0,), (9,)),
+        ((3,), (2,), (0,), (1,)),
+        ((9,), (2,), (2,), (2,)),
+        ((9, 9), (2, 1), None, None),
+        ((9, 9), (4, 1), None, None),
+    ],
+)
+@pytest.mark.parametrize("order", ["lexicographic"])
+@pytest.mark.parametrize("trim_excess", [True, False])
+def test_iter_regions(
+    domain_shape: tuple[int, ...],
+    region_shape: tuple[int, ...],
+    origin: tuple[int, ...] | None,
+    selection_shape: tuple[int, ...] | None,
+    order: _ArrayIndexingOrder,
+    trim_excess: bool,
+) -> None:
+    """
+    Test that iter_regions properly iterates over contiguous regions of a gridded domain.
+    """
+    expected_slices_by_dim: list[list[slice]] = []
+    origin_parsed: tuple[int, ...]
+    selection_shape_parsed: tuple[int, ...]
+    if origin is None:
+        origin_parsed = (0,) * len(domain_shape)
+    else:
+        origin_parsed = origin
+    if selection_shape is None:
+        selection_shape_parsed = tuple(
+            ceildiv(ds, rs) - o
+            for ds, o, rs in zip(domain_shape, origin_parsed, region_shape, strict=True)
+        )
+    else:
+        selection_shape_parsed = selection_shape
+    for d_s, r_s, o, ss in zip(
```

Comment:
Again here this is reimplementing the logic being tested.

Would be good to add some property tests in the future, for example,
1. I think we know what the last index of the last slice should be, depending on `trim_excess` being True or False.
2. we never receive a slice with `start` < corresponding index in origin.
3. `output shape == selection_shape` (with some trim_excess dependency perhaps)
4.